### ISSUE TITLE
ZF 2.4.0 form validation BC break

### ISSUE THREAD

**Issue by AlexeyKosov** [zendframework/zf2#7445]
I don't see any documentation notes about form validation BC break.
Here's the case:

        $form = new \Zend\Form\Form();
        $form->add(array(
            'type' => 'Text',
            'name' => 'email',
        ));
        $form->getInputFilter()->add(array(
            'name' => 'email',
            'required' => false,
            'allow_empty' => false,
            'continue_if_empty' => false,
        ));

        $data = array(
            'foo' => 'bar',
        );

        $form->setData($data);
        if ($form->isValid()) {
            echo "Valid\n";
        } else {
            echo "Invalid\n";
            var_dump($form->getMessages());
        }

Output:

ZF 2.3.7
```
   Valid
```
ZF 2.4.0
```
   Invalid

   array (size=1)
     'email' => 
       array (size=1)
         'isEmpty' => string 'Value is required and can't be empty' (length=36)
```

**Comment by Martin-P** [comment]
This break is caused by PR #4567. The exact issue you are experiencing was already mentioned in the comments in that PR: https://github.com/zendframework/zf2/pull/4567#discussion_r4659825 

> That would allow this to work -- but it also means that order of operations becomes critical -- you would be required to set the "allow empty" flag *after* setting the required flag if you wanted to set it to `false`.

Unfortunatelly it was only said in words it was fixed, but no tests have been added to actually show it was fixed :disappointed: 

##### isValid() returns `false`:
```php
$input = new \Zend\InputFilter\Input();
$input->setName('email')
      ->setRequired(false)
      ->setAllowEmpty(false)
      ->setContinueIfEmpty(false);
var_dump($input->isValid()); // false
```

##### different ordering and now isValid() returns `true`:
```php
$input = new \Zend\InputFilter\Input();
$input->setName('email')
      ->setAllowEmpty(false)
      ->setContinueIfEmpty(false)
      ->setRequired(false);
var_dump($input->isValid()); // true
```

**Comment by Pittiplatsch** [comment]
@Martin-P From https://github.com/zendframework/zf2/pull/4567#issuecomment-19606897 (the answer to your quoted comment):
>I believe the correct way to solve is [...] to set "allow empty" flag to the reverse of "required" by default. In fact, the factory was already doing this if the user didn't explicitly specify allow_empty.
>
>I've updated the factory too to ensure that regardless of the order of the specification, the allow_empty flag specified by the user (if any) is always respected..

So, at first glance, order of specifications *should not* matter.

**Comment by Martin-P** [comment]
@Pittiplatsch The problem is exactly the quote you are referring to. By explicitely setting `allow_empty` in the `setRequired()` method it creates the ordering problem. I fixed this in `Zend\InputFilter\Input`, but it also needs fixing in `Zend\InputFilter\Factory`.

Some specific parameters can only be set at the end when all parameters have been set. That way we avoid getting dependent on the ordering of parameters.

**Comment by weierophinney** [comment]
This is fixed with #7474.
